File: mycollab-web/src/main/java/com/esofthead/mycollab/vaadin/ui/ThemeManager.java
/**
 * This file is part of mycollab-web.
 *
 * mycollab-web is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * mycollab-web is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with mycollab-web.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.esofthead.mycollab.vaadin.ui;

import com.esofthead.mycollab.core.UserInvalidInputException;
import com.esofthead.mycollab.module.user.domain.AccountTheme;
import com.esofthead.mycollab.module.user.service.AccountThemeService;
import com.esofthead.mycollab.spring.ApplicationContextUtil;
import com.esofthead.mycollab.vaadin.AppContext;
import com.vaadin.server.Page;

/**
 * @author MyCollab Ltd.
 * @since 4.1.2
 */
public class ThemeManager {

    public static void loadUserTheme(int sAccountId) {
        AccountThemeService themeService = ApplicationContextUtil.getSpringBean(AccountThemeService.class);

        AccountTheme accountTheme = themeService.findTheme(sAccountId);

        if (accountTheme == null) {
            accountTheme = themeService.findDefaultTheme(AppContext.getAccountId());
            if (accountTheme == null) {
                throw new UserInvalidInputException("Can not load theme for this account. You may get bad experience " +
                        "while using MyCollab. Please contact your site admoinistrator to solve this issue or fill a " +
                        "support request to MyCollab team" +
                        " ");
            }
        }

        StringBuilder extraStyles = new StringBuilder();
		/* Top Menu */
        if (accountTheme.getTopmenubg() != null) {
            extraStyles.append(".topNavigation { background-color: #" + accountTheme.getTopmenubg() + "; }");
        }

        if (accountTheme.getTopmenubgselected() != null) {
            extraStyles.append(".topNavigation .serviceMenuContainer .service-menu.v-buttongroup .v-button.selected { background-color: #"
                    + accountTheme.getTopmenubgselected() + "; }");

            extraStyles.append(".topNavigation .serviceMenuContainer .service-menu.v-buttongroup .v-button:hover { background-color: #"
                    + accountTheme.getTopmenubgselected() + "; }");

            extraStyles.append(".v-button.add-btn-popup:hover { background-color: #" + accountTheme.getTopmenubgselected() + "; }");

        }

        if (accountTheme.getTopmenutext() != null) {
            extraStyles.append(".topNavigation .v-button { color: #" + accountTheme.getTopmenutext() + "; }");
        }

        if (accountTheme.getTopmenutextselected() != null) {
            extraStyles.append(".topNavigation .serviceMenuContainer .service-menu.v-buttongroup .v-button.selected { color: #"
                    + accountTheme.getTopmenutextselected() + "; }");

            extraStyles.append(".topNavigation .serviceMenuContainer .service-menu.v-buttongroup .v-button:hover { color: #"
                    + accountTheme.getTopmenutextselected() + "; }");

            extraStyles.append(".v-button.add-btn-popup:hover { color: #"
                    + accountTheme.getTopmenutextselected() + "; }");
        }

		/* Vertical Tabsheet */

        if (accountTheme.getVtabsheetbg() != null) {
            extraStyles.append(".vertical-tabsheet .navigator-wrap { background-color: #"
                    + accountTheme.getVtabsheetbg() + "; }");
        }

        if (accountTheme.getVtabsheetbgselected() != null) {
            extraStyles.append(".v-progressbar-indicator { background-color: #"
                    + accountTheme.getVtabsheetbgselected() + "; }");

            extraStyles.append(".v-progressbar.medium .v-progressbar-indicator { background-color: #"
                    + accountTheme.getVtabsheetbgselected() + "; }");

            extraStyles.append(".vertical-tabsheet .v-button-tab.tab-selected > .v-button-wrap { background-color: #"
                    + accountTheme.getVtabsheetbgselected() + "; }");

            extraStyles.append(".sidebar-menu .v-button:hover {background-color: #" + accountTheme
                    .getVtabsheetbgselected() + ";}");
        }

        if (accountTheme.getVtabsheettext() != null) {
            extraStyles.append(".vertical-tabsheet .v-button-tab > .v-button-wrap { color: #"
                    + accountTheme.getVtabsheettext() + "; }");
        }

        if (accountTheme.getVtabsheettextselected() != null) {
            extraStyles.append(".vertical-tabsheet .v-button-tab.tab-selected > .v-button-wrap { color: #"
                    + accountTheme.getVtabsheettextselected() + "; }");

            //Color while hover on sidebar menu
            extraStyles.append(".vertical-tabsheet .v-button-tab .v-button-wrap:hover {color: #" + accountTheme
                    .getVtabsheettextselected() + "!important;}");
        }

		/* Tabsheet */
        if (accountTheme.getTabsheetbg() != null) {
            extraStyles.append(".tab-style3 > .v-tabsheet-tabcontainer > .v-tabsheet-tabs > tbody > tr > .v-tabsheet-tabitemcell > " +
                    ".v-tabsheet-tabitem { background-color: #"
                    + accountTheme.getTabsheetbg() + "; }");
        }

        if (accountTheme.getTabsheetbgselected() != null) {
            extraStyles.append(".project-info { background-color: #"
                    + accountTheme.getTabsheetbgselected() + "; }");


            extraStyles.append(".h-sidebar-menu, .projectfeed-hdr-wrapper { background-color: #"
                    + accountTheme.getTabsheetbgselected() + "; }");

            //Set style of popup content
            extraStyles.append(".optionPopupContent .action-wrap .v-button-action { color: #" + accountTheme
                    .getTabsheetbgselected() + "; }");

            //Set link style
            extraStyles.append(".v-app a { color: #" + accountTheme.getTabsheetbgselected() + "; }");

            // Set button link
            extraStyles.append(".v-button.v-button-link { color: #" + accountTheme
                    .getTabsheetbgselected() + "; }");

            extraStyles.append(".tab-style3 > .v-tabsheet-tabcontainer > .v-tabsheet-tabs > tbody > tr >" +
                    " .v-tabsheet-tabitemcell.v-tabsheet-tabitemcell-selected > .v-tabsheet-tabitem { " +
                    "background-color: #" + accountTheme.getTabsheetbgselected() + "; }");

            extraStyles.append(".tab-style3 > .v-tabsheet-tabcontainer > .v-tabsheet-tabs > tbody > tr > " +
                    ".v-tabsheet-tabitemcell > .v-tabsheet-tabitem:hover {background-color: #" +
                    accountTheme.getTabsheetbgselected() + ";}");

            //Set milestone header background
            extraStyles.append(".milestone-view-header {background-color: #" + accountTheme.getTabsheetbgselected() + ";}");

            //Volume display bar in file manager
            extraStyles.append(".v-label.volumeUsageInfo { border-top: 25px solid #" + accountTheme
                    .getTabsheetbgselected() + ";}");
        }

        if (accountTheme.getTabsheettext() != null) {
            extraStyles.append(".tab-style3 > .v-tabsheet-tabcontainer > .v-tabsheet-tabs > tbody > tr > .v-tabsheet-tabitemcell > .v-tabsheet-tabitem { color: #"
                    + accountTheme.getTabsheettext() + "; }");
        }

        if (accountTheme.getTabsheettextselected() != null) {
            extraStyles.append(".tab-style3 > .v-tabsheet-tabcontainer > .v-tabsheet-tabs > tbody > tr > .v-tabsheet-tabitemcell > .v-tabsheet-tabitem.v-tabsheet-tabitem-selected { color: #"
                    + accountTheme.getTabsheettextselected() + "; }");

            extraStyles.append(".tab-style3 > .v-tabsheet-tabcontainer > .v-tabsheet-tabs > tbody > tr > " +
                    ".v-tabsheet-tabitemcell > .v-tabsheet-tabitem:hover {color: #" +
                    accountTheme.getTabsheettextselected() + ";}");

            //Set milestone header text color
            extraStyles.append(".milestone-view-header {color: #" + accountTheme.getTabsheettextselected() + ";}");

            //Volume text display bar in file manager
            extraStyles.append(".v-label.volumeUsageInfo div { color: #" + accountTheme
                    .getTabsheettextselected() + ";}");
        }

		/* Action Buttons */

        if (accountTheme.getActionbtn() != null) {
            extraStyles.append(".v-button.v-button-greenbtn, .v-button-greenbtn:focus { background-color: #" + accountTheme.getActionbtn()
                    + "; border-color: #" + getBorderColor(accountTheme.getActionbtn()) + "; }");

            extraStyles.append(".v-button.add-project-btn {background-color: #" + accountTheme
                    .getActionbtn() + ";}");

            extraStyles.append(".upload-field .v-upload-immediate .v-button {background-color: #" +
                    accountTheme.getActionbtn() + ";}");

            extraStyles.append(".optionPopupContent .action-wrap:hover {" +
                    "background-color: #" + accountTheme.getActionbtn() + "};");

            extraStyles.append(".v-buttongroup.toggle-btn-group .v-button.active { background-color: #" + accountTheme.getActionbtn()
                    + "; border-color: #" + getBorderColor(accountTheme.getActionbtn()) + "; }");

            //Button paging
            extraStyles.append("div.v-button-link.buttonPaging.current, div.v-button-link" +
                    ".buttonPaging:hover { background-color:#" + accountTheme.getActionbtn() + "; }");

            //Selection background of selected item
            extraStyles.append(".v-filterselect-suggestpopup .gwt-MenuItem-selected { background-color:#" +
                    accountTheme.getActionbtn() + "; }");

            //Year block of activity stream
            extraStyles.append(".v-label.year-lbl { box-shadow: 0 0 0 5px #" + accountTheme
                    .getActionbtn() + ";}");

            //Date label of activity stream
            extraStyles.append(".feed-block-wrap .date-lbl { background-color:#" + accountTheme
                    .getActionbtn() + ";}");

            extraStyles.append(".feed-block-wrap .date-lbl::after{ border-left-color:#" + accountTheme
                    .getActionbtn() + ";}");

            // Add style for tree selected
            extraStyles.append("div.v-tree-node-selected {background-color:#" + accountTheme
                    .getActionbtn() + ";}");

            extraStyles.append("div.v-tree-node-selected span {background-color:#" + accountTheme
                    .getActionbtn() + ";}");

            // Button group default button
            extraStyles.append(".v-buttongroup.toggle-btn-group .v-button.btn-group-default {background-color:#" + accountTheme
                    .getActionbtn() + ";}");
        }

        if (accountTheme.getActionbtntext() != null) {
            extraStyles.append(".v-button.v-button-greenbtn, .v-button-greenbtn:focus { color: #"
                    + accountTheme.getActionbtntext() + "; }");

            extraStyles.append(".optionPopupContent .action-wrap .v-button-action .v-button-wrap:hover" +
                    " {" + "color: #" + accountTheme.getActionbtntext() + "};");

            //Button paging
            extraStyles.append("div.v-button-link.buttonPaging.current, div.v-button-link" +
                    ".buttonPaging:hover { color:#" + accountTheme.getActionbtntext() + "; }");

            //Selection text color of selected item
            extraStyles.append(".v-filterselect-suggestpopup .gwt-MenuItem-selected { color:#" +
                    accountTheme.getActionbtntext() + "; }");

            //Date label of activity stream
            extraStyles.append(".feed-block-wrap .date-lbl { color:#" + accountTheme
                    .getActionbtntext() + ";}");

            //Style for tree
            extraStyles.append("div.v-tree-node-selected span {color:#" + accountTheme
                    .getActionbtntext() + ";}");
        }

		/* Option Buttons */

        if (accountTheme.getOptionbtn() != null) {
            extraStyles.append(".v-button.v-button-graybtn, .v-button-graybtn:focus { background-color: #"
                    + accountTheme.getOptionbtn() + "; border-color: #" + getBorderColor(accountTheme.getOptionbtn())
                    + "; }");

            //Set toogle button group background
            extraStyles.append(".v-buttongroup.toggle-btn-group .v-button { background-color: #" +
                    accountTheme.getOptionbtn() + "; border-color: #" + getBorderColor(accountTheme.getOptionbtn()) + ";}");
        }

        if (accountTheme.getOptionbtntext() != null) {
            extraStyles.append(".v-button.v-button-graybtn, .v-button-graybtn:focus { color: #" + accountTheme.getOptionbtntext() + "; }");

            extraStyles.append(".v-buttongroup.toggle-btn-group .v-button { color: #"
                    + accountTheme.getOptionbtntext() + "; }");
        }

		/* Danger Buttons */

        if (accountTheme.getDangerbtn() != null) {
            extraStyles.append(".v-button.v-button-redbtn, .v-button-redbtn:focus { background-color: #"
                            + accountTheme.getDangerbtn() + "; border-color: #"
                            + getBorderColor(accountTheme.getDangerbtn()) + "; }");
        }

        if (accountTheme.getDangerbtntext() != null) {
            extraStyles.append(".v-button.v-button-redbtn, .v-button-redbtn:focus { color: #"
                    + accountTheme.getDangerbtntext() + "; }");
        }

        if (extraStyles.length() > 0) {
            Page.getCurrent().getStyles().add(extraStyles.toString());
        }

    }

    public static void loadDemoTheme(AccountTheme accountTheme) {
        StringBuilder demoExtraStyles = new StringBuilder();
		/* Top Menu */

        if (accountTheme.getTopmenubg() != null) {
            demoExtraStyles.append(".example-block .topNavigation { background-color: #"
                    + accountTheme.getTopmenubg() + "; }");
        }

        if (accountTheme.getTopmenubgselected() != null) {
            demoExtraStyles.append(".example-block .topNavigation .service-menu.v-buttongroup .v-button.selected { background-color: #"
                    + accountTheme.getTopmenubgselected() + "; }");
        }

        if (accountTheme.getTopmenutext() != null) {
            demoExtraStyles.append(".example-block .topNavigation .v-button-caption { color: #"
                    + accountTheme.getTopmenutext() + "; }");
        }

        if (accountTheme.getTopmenutextselected() != null) {
            demoExtraStyles.append(".example-block .topNavigation .service-menu.v-buttongroup .v-button.selected .v-button-caption { color: #"
                    + accountTheme.getTopmenutextselected() + "; }");
        }

		/* Vertical Tabsheet */

        if (accountTheme.getVtabsheetbg() != null) {
            demoExtraStyles.append(".example-block .navigator-wrap { background-color: #"
                    + accountTheme.getVtabsheetbg() + "; }");
        }

        if (accountTheme.getVtabsheetbgselected() != null) {
            demoExtraStyles.append(".example-block .vertical-tabsheet .v-button-tab.tab-selected > .v-button-wrap { background-color: #"
                    + accountTheme.getVtabsheetbgselected() + "; }");
        }

        if (accountTheme.getVtabsheettext() != null) {
            demoExtraStyles.append(".example-block .vertical-tabsheet .v-button-tab > .v-button-wrap > .v-button-caption { color: #"
                    + accountTheme.getVtabsheettext() + "; }");
        }

        if (accountTheme.getVtabsheettextselected() != null) {
            demoExtraStyles.append(".example-block .vertical-tabsheet .v-button-tab.tab-selected > .v-button-wrap > .v-button-caption { color: #"
                    + accountTheme.getVtabsheettextselected() + "; }");
        }

		/* Tabsheet */

        if (accountTheme.getTabsheetbg() != null) {
            demoExtraStyles.append(".example-block .tab-style3 > .v-tabsheet-tabcontainer > .v-tabsheet-tabs > tbody > tr > .v-tabsheet-tabitemcell > .v-tabsheet-tabitem { background-color: #"
                    + accountTheme.getTabsheetbg() + "; }");
        }

        if (accountTheme.getTabsheetbgselected() != null) {
            demoExtraStyles.append(".example-block .tab-style3 > .v-tabsheet-tabcontainer > .v-tabsheet-tabs > tbody > tr > .v-tabsheet-tabitemcell.v-tabsheet-tabitemcell-selected > .v-tabsheet-tabitem { background-color: #"
                    + accountTheme.getTabsheetbgselected() + "; }");
        }

        if (accountTheme.getTabsheettext() != null) {
            demoExtraStyles.append(".example-block .tab-style3 > .v-tabsheet-tabcontainer > .v-tabsheet-tabs > tbody > tr > .v-tabsheet-tabitemcell > .v-tabsheet-tabitem .v-caption .v-captiontext { color: #"
                            + accountTheme.getTabsheettext() + "; }");
        }

        if (accountTheme.getTabsheettextselected() != null) {
            demoExtraStyles.append(".example-block .tab-style3 > .v-tabsheet-tabcontainer > .v-tabsheet-tabs > tbody > tr > .v-tabsheet-tabitemcell > .v-tabsheet-tabitem.v-tabsheet-tabitem-selected .v-caption .v-captiontext { color: #"
                    + accountTheme.getTabsheettextselected() + "; }");
        }

		/* Action Buttons */

        if (accountTheme.getActionbtn() != null) {
            demoExtraStyles.append(".example-block .v-button.v-button-greenbtn, .example-block .v-button-greenbtn:focus { background-color: #"
                    + accountTheme.getActionbtn()
                    + "; border-color: #"
                    + getBorderColor(accountTheme.getActionbtn())
                    + "; }");
        }

        if (accountTheme.getActionbtntext() != null) {
            demoExtraStyles.append(".example-block .v-button.v-button-greenbtn, .example-block .v-button-greenbtn:focus { color: #"
                    + accountTheme.getActionbtntext() + "; }");
        }

		/* Option Buttons */

        if (accountTheme.getOptionbtn() != null) {
            demoExtraStyles.append(".example-block .v-button.v-button-graybtn, .example-block .v-button-graybtn:focus { background-color: #"
                    + accountTheme.getOptionbtn()
                    + "; border-color: #"
                    + getBorderColor(accountTheme.getOptionbtn())
                    + "; }");
        }

        if (accountTheme.getOptionbtntext() != null) {
            demoExtraStyles.append(".example-block .v-button.v-button-graybtn, .example-block .v-button-graybtn:focus { color: #"
                    + accountTheme.getOptionbtntext() + "; }");
        }

		/* Danger Buttons */

        if (accountTheme.getDangerbtn() != null) {
            demoExtraStyles.append(".example-block .v-button.v-button-redbtn, .example-block .v-button-redbtn:focus { background-color: #"
                    + accountTheme.getDangerbtn()
                    + "; border-color: #"
                    + getBorderColor(accountTheme.getDangerbtn())
                    + "; }");
        }

        if (accountTheme.getDangerbtntext() != null) {
            demoExtraStyles.append(".example-block .v-button.v-button-redbtn, .example-block .v-button-redbtn:focus { color: #"
                            + accountTheme.getDangerbtntext() + "; }");
        }

        if (demoExtraStyles.length() > 0) {
            Page.getCurrent().getStyles().add(demoExtraStyles.toString());
        }

    }

    private static String getBorderColor(String bgColor) {
        String rgb = "";
        Long c;
        for (int i = 0; i < 3; i++) {
            c = Long.valueOf(bgColor.substring(i * 2, i * 2 + 2), 16);
            c = Math.round(Math.min(Math.max(0, c - c * 0.2), 255));
            String strVal = Long.toHexString(c);
            rgb += ("00" + strVal).substring(strVal.length());
        }

        return rgb;
    }

}

File: mycollab-ui/src/main/java/com/mycollab/vaadin/ui/ThemeManager.kt
/**
 * Copyright © MyCollab
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http:></http:>//www.gnu.org/licenses/>.
 */
package com.mycollab.vaadin.ui

import com.mycollab.common.i18n.ShellI18nEnum
import com.mycollab.core.UserInvalidInputException
import com.mycollab.core.utils.ColorUtils
import com.mycollab.module.user.domain.AccountTheme
import com.mycollab.module.user.service.AccountThemeService
import com.mycollab.spring.AppContextUtil
import com.mycollab.vaadin.AppUI
import com.mycollab.vaadin.UserUIContext
import com.vaadin.server.Page

/**
 * @author MyCollab Ltd.
 * @since 4.1.2
 */
object ThemeManager {

    @JvmStatic
    fun loadDesktopTheme(sAccountId: Int) {
        val themeService = AppContextUtil.getSpringBean(AccountThemeService::class.java)
        var accountTheme = themeService.findTheme(sAccountId)

        if (accountTheme == null) {
            accountTheme = themeService.findDefaultTheme(AppUI.accountId)
            if (accountTheme == null) {
                throw UserInvalidInputException(UserUIContext.getMessage(ShellI18nEnum.ERROR_CAN_NOT_LOAD_THEME))
            }
        }

        val extraStyles = StringBuilder()
        /* Top Menu */
        if (accountTheme.topmenubg != null) {
            extraStyles.append(".topNavigation { background-color: #${accountTheme.topmenubg}; }")
            extraStyles.append("#login-header { background-color: #${accountTheme.topmenubg}; }")
            extraStyles.append(".topNavigation #mainLogo { background-color: #${accountTheme.topmenubg}; }")
        }

        if (accountTheme.topmenubgselected != null) {
            extraStyles.append(".topNavigation .serviceMenuContainer .service-menu .v-button.selected { background-color: #${accountTheme.topmenubgselected}; }")
            extraStyles.append(".topNavigation .serviceMenuContainer .service-menu .v-button:hover { background-color: #${accountTheme.topmenubgselected}; }")
            extraStyles.append(".v-button.add-btn-popup:hover { background-color: #${accountTheme.topmenubgselected}; }")
            extraStyles.append(".topNavigation .v-button.ad { background-color: #${accountTheme.topmenubgselected}; }")
        }

        if (accountTheme.topmenutext != null) {
            extraStyles.append(".topNavigation .v-button { color: #${accountTheme.topmenutext}; }")
            extraStyles.append(".subDomain { color: #${accountTheme.topmenutext}; }")
            extraStyles.append(".accountMenuContainer .v-popup-indicator::before { color: #${accountTheme.topmenutext}; }")
        }

        if (accountTheme.topmenutextselected != null) {
            extraStyles.append(".topNavigation .serviceMenuContainer .service-menu .v-button.selected { color: #${accountTheme.topmenutextselected}; }")
            extraStyles.append(".topNavigation .serviceMenuContainer .service-menu .v-button:hover { color: #${accountTheme.topmenutextselected}; }")
            extraStyles.append(".v-button.add-btn-popup:hover { color: #${accountTheme.topmenutextselected}; }")
            extraStyles.append(".topNavigation .v-button.ad { color: #${accountTheme.topmenutextselected}; }")
            extraStyles.append(".topNavigation .v-button.ad .v-icon { color: #${accountTheme.topmenutextselected}; }")
        }

        /* Vertical Tabsheet */

        if (accountTheme.vtabsheetbg != null) {
            extraStyles.append(".vertical-tabsheet .navigator-wrap { background-color: #${accountTheme.vtabsheetbg}; }")
        }

        if (accountTheme.vtabsheettext != null) {
            extraStyles.append(".vertical-tabsheet .v-button-tab > .v-button-wrap { color: #${accountTheme.vtabsheettext}; }")
            extraStyles.append(".closed-button .v-button-wrap .v-icon { color: #${accountTheme.vtabsheettext}; }")
            extraStyles.append(".expand-button .v-button-wrap .v-icon { color: #${accountTheme.vtabsheettext}; }")
            extraStyles.append(".project-info .header { color: #${accountTheme.vtabsheettext}; }")
            extraStyles.append(".intro-text-wrap .v-label { color: #${accountTheme.vtabsheettext}; }")
        }

        if (accountTheme.vtabsheetbgselected != null) {
            extraStyles.append(".vertical-tabsheet .v-button-tab.tab-selected { background-color: #${accountTheme.vtabsheetbgselected}; }")
            extraStyles.append(".vertical-tabsheet .v-button-tab:hover {background-color: #${accountTheme.vtabsheetbgselected};}")
        }

        if (accountTheme.vtabsheettextselected != null) {
            extraStyles.append(".vertical-tabsheet .v-button-tab.tab-selected > .v-button-wrap { color: #${accountTheme.vtabsheettextselected}; }")
            //Color while hover on sidebar menu
            extraStyles.append(".vertical-tabsheet .v-button-tab .v-button-wrap:hover {color: #${accountTheme.vtabsheettextselected}!important;}")
            extraStyles.append(".vertical-tabsheet .v-button-tab:hover .v-button-wrap {color: #${accountTheme.vtabsheettextselected}!important;}")
            //Volume text display bar in file manager
            extraStyles.append(".v-label.volumeUsageInfo div { color: #${accountTheme.vtabsheettextselected};}")
        }

        /* Action Buttons */

        if (accountTheme.actionbtn != null) {
            extraStyles.append(".v-button.v-button-greenbtn, .v-button-greenbtn:focus { background-color: #${accountTheme.actionbtn}; }")
            extraStyles.append(".splitbutton:hover .v-button.v-button-greenbtn, .v-button-greenbtn:hover { background-color: ${ColorUtils.darkerColor("#" + accountTheme.actionbtn)}; }")
            extraStyles.append(".upload-field .v-upload-immediate .v-button {background-color: #${accountTheme.actionbtn};}")
            extraStyles.append(".upload-field .v-upload-immediate .v-button:hover {background-color: ${ColorUtils.darkerColor("#" + accountTheme.actionbtn)};}")
            extraStyles.append(".optionPopupContent .action-wrap:hover {background-color: #${accountTheme.actionbtn}};")
            extraStyles.append(".v-buttongroup.toggle-btn-group .v-button.active { background-color: #${accountTheme.actionbtn}; }")
            //Button paging
            extraStyles.append(".v-button.buttonPaging.current, .v-button.buttonPaging:hover { background-color:#${accountTheme.actionbtn}; }")
            //Selection background of selected item
            extraStyles.append(".v-filterselect-suggestpopup .gwt-MenuItem-selected { background-color:#${accountTheme.actionbtn}; }")
            //Year block of activity stream
            extraStyles.append(".v-label.year-lbl { box-shadow: 0 0 0 5px #${accountTheme.actionbtn};}")
            //Date label of activity stream
            extraStyles.append(".activity-list .feed-block-wrap .date-lbl { background-color:#${accountTheme.actionbtn};}")
            extraStyles.append(".activity-list .feed-block-wrap .date-lbl::after{ border-left-color:#${accountTheme.actionbtn};}")
            extraStyles.append(".activity-list .feed-block-wrap:hover .date-lbl { background-color:" + ColorUtils.darkerColor("#" + accountTheme.actionbtn) + ";}")
            extraStyles.append(".activity-list .feed-block-wrap:hover .date-lbl::after{ border-left-color:${ColorUtils.darkerColor("#" + accountTheme.actionbtn)};}")

            // Button group default button
            extraStyles.append(".toggle-btn-group .v-button.btn-group-default {background-color:#${accountTheme.actionbtn};}")
            extraStyles.append(".toggle-btn-group .v-button.btn-group-default:hover {background-color:${ColorUtils.darkerColor("#" + accountTheme.actionbtn)};}")
            extraStyles.append(".v-context-menu-container .v-context-menu .v-context-submenu:hover {background-color:#${accountTheme.actionbtn};}")
        }

        if (accountTheme.actionbtntext != null) {
            extraStyles.append(".v-button.v-button-greenbtn, .v-button-greenbtn:focus { color: #${accountTheme.actionbtntext}; }")

            extraStyles.append(".upload-field .v-upload-immediate .v-button, .upload-field .v-upload-immediate .v-button:focus {color: #${accountTheme.actionbtntext};}")

            extraStyles.append(".optionPopupContent .action-wrap .v-button-action .v-button-wrap:hover {color: #${accountTheme.actionbtntext}};")

            //Button paging
            extraStyles.append(".v-button.buttonPaging.current, .v-button.buttonPaging:hover { color:#${accountTheme.actionbtntext}; }")

            //Selection text color of selected item
            extraStyles.append(".v-filterselect-suggestpopup .gwt-MenuItem-selected { color:#${accountTheme.actionbtntext}; }")

            //Date label of activity stream
            extraStyles.append(".activity-list .feed-block-wrap .date-lbl { color:#${accountTheme.actionbtntext};}")

            extraStyles.append(".v-button.v-button-block {color:#${accountTheme.actionbtntext};}")

            extraStyles.append(".v-context-menu-container .v-context-menu .v-context-submenu:hover {color:#${accountTheme.actionbtntext};}")

            extraStyles.append(".toggle-btn-group .v-button.btn-group-default {color:#${accountTheme.actionbtntext};}")
        }

        /* Option Buttons */

        if (accountTheme.optionbtn != null) {
            extraStyles.append(".v-button.v-button-graybtn, .v-button-graybtn:focus { background-color: #${accountTheme.optionbtn};}")
            extraStyles.append(".splitbutton:hover .v-button-graybtn, .v-button-graybtn:hover { background-color: ${ColorUtils.darkerColor("#" + accountTheme.optionbtn)};}")
            //Set toggle button group background
            extraStyles.append(".toggle-btn-group .v-button { background-color: #${accountTheme.optionbtn};}")
            extraStyles.append(".toggle-btn-group .v-button:hover { background-color: ${ColorUtils.darkerColor("#" + accountTheme.optionbtn)};}")
        }

        if (accountTheme.optionbtntext != null) {
            extraStyles.append(".v-button.v-button-graybtn, .v-button-graybtn:focus { color: #${accountTheme.optionbtntext}; }")
            extraStyles.append(".toggle-btn-group .v-button { color: #${accountTheme.optionbtntext}; }")
        }

        /* Danger Buttons */

        if (accountTheme.dangerbtn != null) {
            extraStyles.append(".v-button.v-button-redbtn, .v-button-redbtn:focus { background-color: #${accountTheme.dangerbtn}; }")
            extraStyles.append(".v-button-redbtn:hover { background-color: ${ColorUtils.darkerColor("#" + accountTheme.dangerbtn, 0.1)}; }")
            //Set style of popup content action
            extraStyles.append(".optionPopupContent .action-wrap.danger .v-button-action { color: #${accountTheme.dangerbtn}; }")
            extraStyles.append(".optionPopupContent .action-wrap.danger:hover {background-color: #${accountTheme.dangerbtn};}")
        }

        if (accountTheme.dangerbtntext != null) {
            extraStyles.append(".v-button.v-button-redbtn, .v-button-redbtn:focus { color: #${accountTheme.dangerbtntext}; }")
        }

        if (extraStyles.isNotBlank()) {
            Page.getCurrent().styles.add(extraStyles.toString())
        }

    }

    @JvmStatic
    fun loadDemoTheme(accountTheme: AccountTheme) {
        val demoExtraStyles = StringBuilder()
        /* Top Menu */

        if (accountTheme.topmenubg != null) {
            demoExtraStyles.append(".example-block .topNavigation { background-color: #${accountTheme.topmenubg}; }")
        }

        if (accountTheme.topmenubgselected != null) {
            demoExtraStyles.append(".example-block .topNavigation .service-menu.v-buttongroup .v-button.selected { background-color: #${accountTheme.topmenubgselected}; }")
        }

        if (accountTheme.topmenutext != null) {
            demoExtraStyles.append(".example-block .topNavigation .v-button-caption { color: #${accountTheme.topmenutext}; }")
        }

        if (accountTheme.topmenutextselected != null) {
            demoExtraStyles.append(".example-block .topNavigation .service-menu.v-buttongroup .v-button.selected .v-button-caption { color: #${accountTheme.topmenutextselected}; }")
        }

        /* Vertical Tabsheet */
        if (accountTheme.vtabsheetbg != null) {
            demoExtraStyles.append(".example-block .navigator-wrap { background-color: #${accountTheme.vtabsheetbg}; }")
        }

        if (accountTheme.vtabsheetbgselected != null) {
            demoExtraStyles.append(".example-block .vertical-tabsheet .v-button-tab.tab-selected { background-color: #${accountTheme.vtabsheetbgselected}; }")
        }

        if (accountTheme.vtabsheettext != null) {
            demoExtraStyles.append(".example-block .vertical-tabsheet .v-button-tab > .v-button-wrap > .v-button-caption { color: #${accountTheme.vtabsheettext}; }")
        }

        if (accountTheme.vtabsheettextselected != null) {
            demoExtraStyles.append(".example-block .vertical-tabsheet .v-button-tab.tab-selected > .v-button-wrap > .v-button-caption { color: #${accountTheme.vtabsheettextselected}; }")
        }

        /* Action Buttons */
        if (accountTheme.actionbtn != null) {
            demoExtraStyles.append(".example-block .v-button.v-button-greenbtn, .example-block .v-button-greenbtn:focus { background-color: #${accountTheme.actionbtn}; }")
        }

        if (accountTheme.actionbtntext != null) {
            demoExtraStyles.append(".example-block .v-button.v-button-greenbtn, .example-block .v-button-greenbtn:focus { color: #${accountTheme.actionbtntext}; }")
        }

        /* Option Buttons */

        if (accountTheme.optionbtn != null) {
            demoExtraStyles.append(".example-block .v-button.v-button-graybtn, .example-block .v-button-graybtn:focus { background-color: #${accountTheme.optionbtn}; }")
        }

        if (accountTheme.optionbtntext != null) {
            demoExtraStyles.append(".example-block .v-button.v-button-graybtn, .example-block .v-button-graybtn:focus { color: #${accountTheme.optionbtntext}; }")
        }

        /* Danger Buttons */
        if (accountTheme.dangerbtn != null) {
            demoExtraStyles.append(".example-block .v-button.v-button-redbtn, .example-block .v-button-redbtn:focus { background-color: #${accountTheme.dangerbtn}; }")
        }

        if (accountTheme.dangerbtntext != null) {
            demoExtraStyles.append(".example-block .v-button.v-button-redbtn, .example-block .v-button-redbtn:focus { color: #${accountTheme.dangerbtntext}; }")
        }

        if (demoExtraStyles.isNotBlank()) {
            Page.getCurrent().styles.add(demoExtraStyles.toString())
        }
    }
}
